concrete/src/Updater/Migrations/Migrations/Version20181006212400.php

```diff
@@ -0,0 +1,116 @@
+            $config->set("database.connections.{$connectionName}", $connectionConfig);
+            $config->save("database.connections.{$connectionName}", $connectionConfig);
+        } catch (Exception $x) {
+            $this->output(t('Failed to set character sets: %s', $x->getMessage()));
+        }
+        finally {
+            try {
+                $this->connection->executeQuery('SET foreign_key_checks = 1');
+            } catch (Exception $x) {
+            }
+        }
+    }
+
+    /**
+     * @param \Concrete\Core\Database\Connection\Connection $connection
+     *
+     * @throws \Exception
+     *
+     * @return string[]
+     */
+    protected function getPreferredCharsetCollation(Connection $connection)
+    {
+        $config = $this->app->make('config');
+        $charset = strtolower((string) $config->get('database.preferred_character_set'));
+        if ($charset === '') {
+            throw new Exception(t('no preferred character set defined.'));
+        }
+        $supportedCharsets = $this->connection->getSupportedCharsets();
+        if (!isset($supportedCharsets[$charset])) {
+            throw new Exception(t('the character set "%s" is not supported by the database.', $charset));
+        }
+        $collation = strtolower((string) $config->get('database.preferred_collation'));
+        if ($collation === '' || $collation === $supportedCharsets[$charset]) {
+            $collation = $supportedCharsets[$charset];
+        } else {
+            $supportedCollations = $this->connection->getSupportedCollations();
+            if (!isset($supportedCollations[$collation])) {
+                throw new Exception(t('the collation "%s" is not supported by the database.', $collation));
+            }
+            if ($supportedCollations[$collation] !== $charset) {
+                throw new Exception(t('the collation "%1$s" is associated to the character set "%2%s" and not to "%3%s".', $collation, $supportedCollations[$collation], $charset));
+            }
+        }
+
+        return [$charset, $collation];
+    }
+
+    /**
+     * @param \Concrete\Core\Database\Connection\Connection $connection
+     * @param string $tableName
+     * @param string $charset
+     * @param string $collation
+     *
+     * @throws \Exception
+     *
+     * @return bool true: table updated, false: table already up-to-date
+     */
+    protected function updateTable(Connection $connection, $tableName, $charset, $collation)
+    {
+        $row = $connection->fetchAssoc('SHOW TABLE STATUS WHERE name like ?', [$tableName]);
+        if ($row === false || !isset($row['Collation'])) {
+            throw new Exception(t('failed to retrieve the table collation.'));
+        }
+        if (strcasecmp($collation, $row['Collation']) === 0) {
+            return false;
+        }
+        $connection->executeQuery('ALTER TABLE ' . $connection->quoteIdentifier($tableName) . ' CONVERT TO CHARACTER SET ' . $charset . ' COLLATE ' . $collation);
```

Comment:
Scrutinizer warns that this is potential SQLi since `$charset` and `$collation` come directly from config. Is it worth using prepared statements here?

Reply:
Well, from the [MySQL manual](https://dev.mysql.com/doc/refman/5.5/en/alter-table.html) I read that the charset/collation names are not enclosed in quotes, so we should not use `quote` or anything else.
Furthermore, if someone is able to write harmful stuff to the configuration files, they are also able to gain full control over MySQL and the whole site

Reply:
> I still think it's worth reducing the complexity in a couple of spots here.

I [introduced](https://github.com/concrete5/concrete5/pull/7169/commits/bb2f23376e1c07c37f817c030f48819ece00a799) a helper class to reduce complexity.

> Scrutinizer warns that this is potential SQLi since `$charset` and `$collation` come directly from config. Is it worth using prepared statements here?

Now we check that character sets and collations contain only alphanumeric characters (see [here]/https://github.com/concrete5/concrete5/pull/7169/commits/bb2f23376e1c07c37f817c030f48819ece00a799#diff-5844c15f02f687f6c130df791dd6be8aR165) and [here](https://github.com/concrete5/concrete5/pull/7169/commits/bb2f23376e1c07c37f817c030f48819ece00a799#diff-5844c15f02f687f6c130df791dd6be8aR177).